src/main/java/com/google/cloud/pubsublite/kafka/SinglePartitionSubscriber.java

```diff
@@ -0,0 +1,150 @@
+
+  @GuardedBy("monitor.monitor")
+  private BlockingPullSubscriber subscriber;
+
+  @GuardedBy("monitor.monitor")
+  private boolean needsCommitting = false;
+
+  @GuardedBy("monitor.monitor")
+  private Optional<Offset> lastReceived = Optional.empty();
+
+  SinglePartitionSubscriber(
+      PullSubscriberFactory subscriberFactory,
+      Partition partition,
+      SeekRequest initialSeek,
+      Committer committer)
+      throws CheckedApiException {
+    this.subscriberFactory = subscriberFactory;
+    this.partition = partition;
+    this.committer = committer;
+    this.subscriber =
+        subscriberFactory.newPullSubscriber(partition, initialSeek, this::onSubscriberReset);
+  }
+
+  /** Executes a client-initiated seek. */
+  void clientSeek(SeekRequest request) throws CheckedApiException {
+    try (CloseableMonitor.Hold h = monitor.enter()) {
+      subscriber.close();
+      subscriber = subscriberFactory.newPullSubscriber(partition, request, this::onSubscriberReset);
+    }
+  }
+
+  ApiFuture<Void> onData() {
+    try (CloseableMonitor.Hold h = monitor.enter()) {
+      return subscriber.onData();
+    }
+  }
+
+  @GuardedBy("monitor.monitor")
+  private ArrayDeque<SequencedMessage> pullMessages() throws CheckedApiException {
+    ArrayDeque<SequencedMessage> messages = new ArrayDeque<>();
+    for (Optional<SequencedMessage> message = subscriber.messageIfAvailable();
+        message.isPresent();
+        message = subscriber.messageIfAvailable()) {
+      messages.add(message.get());
+    }
+    return messages;
+  }
+
+  /** Pulls all available messages. */
+  ArrayDeque<SequencedMessage> getMessages() throws CheckedApiException {
+    try (CloseableMonitor.Hold h = monitor.enter()) {
+      ArrayDeque<SequencedMessage> messages = pullMessages();
+      if (!messages.isEmpty()) {
+        lastReceived = Optional.of(Iterables.getLast(messages).offset());
+        needsCommitting = true;
+      }
+      return messages;
+    }
+  }
+
+  Optional<Long> position() {
+    try (CloseableMonitor.Hold h = monitor.enter()) {
+      return lastReceived.map(lastReceived -> lastReceived.value() + 1);
+    }
+  }
+
+  /** Executes a client-initiated commit. */
+  ApiFuture<Void> commitOffset(Offset offset) {
+    return committer.commitOffset(offset);
+  }
+
+  /** Auto-commits the offset of the last received message. */
+  Optional<ApiFuture<Offset>> autoCommit() {
+    try (CloseableMonitor.Hold h = monitor.enter()) {
+      if (!needsCommitting) return Optional.empty();
+      checkState(lastReceived.isPresent());
+      needsCommitting = false;
+      // The Pub/Sub Lite commit offset is one more than the last received.
+      Offset toCommit = Offset.of(lastReceived.get().value() + 1);
+      return Optional.of(
+          ApiFutures.transform(
+              committer.commitOffset(toCommit),
+              ignored -> toCommit,
+              MoreExecutors.directExecutor()));
+    }
+  }
+
+  void close() throws CheckedApiException {
```

Comment:
use ProxyService to implement lifetime management of the committer if you could.

Reply:
Done